[Catsbi][5주차] 5주차 미션 유효성 검사

catsbi opened the pull request:
안녕하세요. 
원래 매 주차 첫째 날은 시간이 퇴근 후 시간이 부족해서 미완성상태로 제출하기가 좀 그래서 자주 패스를 하곤했는데, 이젠 구현을 거의 못했더라도 무조건 제출하도록 하려고 합니다. 

5주차도 잘 부탁드립니다!

catsbi committed:
feat: 상품에 Validation annotation 추가

catsbi committed:
feat: 회원 정보 도메인, 서비스, 리포지토리, 컨트롤러 생성 및 DozerMapper 빈 등록

catsbi committed:
feat: Account 도메인 테스트 케이스 생성

catsbi committed:
feat: Account 서비스, 리포지토리, 컨트롤러 기본 골격 코드 구현및 테스트 클래스 소스 파일 생성

catsbi committed:
refactor: reformat code, 개행 리팩토링

hannut91 commented on app/src/test/java/com/codesoom/assignment/dto/AccountDataTest.java:
```diff
@@ -0,0 +1,45 @@
+import static com.codesoom.assignment.Constant.ACCOUNT_PASSWORD;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("DTO 객체인 AccountData의 비즈니스 로직 테스트")
```
DTO가 하는 일을 비즈니스 로직이라고 불러도 될까요?

catsbi replied:
이 부분에 대해서 뭐라 지칭해야 할지 고민을 했는데, 인터넷에서 DTO의 로직 테스트로 검색을 하다보니 비즈니스 로직이라고 지칭을 하는 곳이 많아서 비즈니스 로직이라 지칭했습니다.

hannut91 replied:
제가 생각하기에는 애플리케이션의 비즈니스 로직과 헷갈려서 비즈니스 로직이라는 말을 여기서 사용하지 않는 것이 좋다고 생각해요. 한솔님의 생각은 어떠신가요?

hannut91 reviewed (commented):
좋습니다 ㅎㅎ 저희는 연습을 하고 있는 것이니 항상 테스트 먼저! 를 기억해주셨으면 좋겠습니다

Commit:
feat: AccountData DTO 유효성 검증 테스트 케이스 추가

Commit:
feat: AccountService 테스트케이스 추가 및 회원 검색 실패 예외 객체 생성

Commit:
feat: AccountService 테스트케이스 추가 및 AccountService 기능 구현

Commit:
feat: AccountController javadoc 작성

Commit:
feat: AccountController 테스트 케이스 추가 및 작성

Commit:
feat: AccountController 식별자 실패 테스트 케이스 추가

Commit:
feat: 회원 정보 업데이트 DTO(AccountUpdateData) 생성및 유효성 로직 별도 추가, 회원 서비스 계층형 테스트 추가, 업데이트 실패 예외 객체 생성

catsbi committed:
refactor: 필요없는 메서드 제거 및 회원 정보 수정 테스트 케이스 추가

catsbi commented:
안녕하세요. 오늘도 잘 부탁드립니다. 

기능을 구현하고 테스트를 하던 도중 몇가지 궁금증이 있습니다. 

1. Lombok에 의해 생성되는 로직(ex: toString, builder, ...)으로 인해 코드 커버리지가 떨어지는데 builder는 생성자 테스트라는 이름으로 테스트 할 수 있지만 toString같은것도 모두 사용해보면서 테스트를 해야하나요? 

2. tests에 있는 step 테스트를 진행하다보니 회원 정보 수정 폼에서 필드가 ""이고 식별자도 9999로 존재하지 않을 경우 NotFound가 아니라 BadRequest가 되야하는데, 그럼 필드의 공백 유효성 검증을 먼저 해줘야 합니다. 

여기서 그렇기에 updateForm 객체에서 name, password, email 필드를 NotBlank 애노테이션으로 검증해주게되면 잘 동작은하지만, 업데이트를 원하지 않는 필드는 값을 전달하지않아 null 인 상황에도 유효성에 걸려서 수정로직이 제대로 작동을 안합니다. 
저는 그래서 이 updateForm의 경우에는 유효성검증 애노테이션을 빼고 메서드로 구현해서 유효성 검증을 해줬습니다. 

혹시 애노테이션으로도 해결이 가능한 문제일까요?

hannut91 commented:
> Lombok에 의해 생성되는 로직(ex: toString, builder, ...)으로 인해 코드 커버리지가 떨어지는데 builder는 생성자 테스트라는 이름으로 테스트 할 수 있지만 toString같은것도 모두 사용해보면서 테스트를 해야하나요?

Lombok에 의해 생성되는 로직은 테스트해도 되고 안해도 됩니다. 테스트를 작성했을 때 무엇을 얻을지에 따라서 다를 것 같아요. 예를들어 Builder에 대해서 아주 자세하게 테스트를 작성하는 것은 Lombok자체에 대한 검증이기 때문에, 우리가 할 필요 없을 것 같아요. 하지만 우리의 코드가 아닌 외부 라이브러리 코드이기 때문에, 외부 라이브러리가 우리가 생각한 의도대로 동작하나? 를 검증하기 위해서라면 테스트를 작성하는 것이 의미가 있을 것 같아요.

Lombok은 사실 너무 잘 알려진 라이브러리고, Builder같은 것들이 동작이 변경될 일이 거의 없을 것이라고 기대하기 때문에 별로 작성하고 싶지 않지만, 만약 검증되지 않았고, 버전이 자주 바뀌는 라이브러리에 대해서 작성하는 것은 의미가 있다고 봅니다. `테스트 주도 개발로 배우는 객체 지향 설계와 실천`에서 이러한 외부 라이브러리를 어떻게 테스트할지에 대해 다루므로 보시면 좋을 것 같습니다.

![KakaoTalk_Photo_2021-09-07-23-55-11](https://user-images.githubusercontent.com/14071105/132366341-50af931f-9edb-4c28-afb4-26924e10ab4b.jpeg)

2. updateForm 객체에서 name, password, email 필드를 NotBlank 애노테이션으로 검증해주게되면 잘 동작은하지만, 업데이트를 원하지 않는 필드는 값을 전달하지않아 null 인 상황에도 유효성에 걸려서 수정로직이 제대로 작동을 안합니다.

NotBlank이면서 null을 허용해야 하는 상황인 것 같아요. 이건 찾아봐야 할 것 같습니다

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+        accountRepository = mock(AccountRepository.class);
+        accountService = new AccountService(accountRepository);
+
+        setUpFixture();
```
fixture는 테스트를 위한 데이터를 나타냅니다. `mockSetUp`이란 이름은 어떨까요?

catsbi replied:
네이밍은 참 어렵네요 😿 mockSetUp이 적절한 이름 같습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceTest.java:
```diff
@@ -0,0 +1,109 @@
+    @DisplayName("회원 정보를 등록할 수 있다. ")
+    @Test
+    void createWithValidAccountV1() {
+        AccountSaveData savedAccount = accountService.creation(accountData);
```
여기 테스트를 보니 DTO를 입력으로 받고 DTO를 출력하고있는 것 같습니다. 이 DTO는 외부 요청에 대한 데이터를 전송하기 위해서 사용되는 것으로 이해했는데, 그렇다면 이 DTO가 변경되었을 때 `AccountService`도 같이 변경될 것 같아요. 그러면 서비스에서 외부 Controller에 대해서 의존을 할 것 같은데 괜찮을까요?

catsbi replied:
해당 리뷰를 보고 DTO의 사용 범위에 대해서 공부를 좀 해봤는데, 이 주제가 이미 많이 언급되어 논의되던 이슈였군요 ㅎㅎ 

기본적으로는 DTO는 컨트롤러에서만 사용하며 서비스계층에는 들어가지 않도록 해야 한다는데, 저는 해당 DTO->도메인(혹은 도메인 -> DTO)으로 변환되기 위해서는 상황에 따라 다른 엔티티를 참조하고 서비스 로직을 사용해서 조합해서 만들어야 할 경우도 있을 수 있기에 이런 변환 작업이 컨트롤러에 있을 경우 컨트롤러에서 여러 서비스에 의존할 수 있기에 문제가 될 수 있다고 생각했습니다. 

그리고 DTO의 내용이 바뀌더라도 도메인으로 변경되는 변환에 대한 책임은 서비스가 아닌 DTO가 가지고 있기에 서비스측에서는 DTO의 내용이 바뀌더라도 괜찮을 것 같습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/controllers/AccountController.java:
```diff
@@ -0,0 +1,78 @@
+    @PatchMapping("{id}")
+    public AccountSaveData patchAccount(@PathVariable long id,
+                                        @RequestBody @Valid AccountUpdateData accountData) {
+        if (!
+                accountData.isValid()) {
+            throw new AccountUpdateFailedException();
+        }
```
```suggestion
        if (!accountData.isValid()) {
            throw new AccountUpdateFailedException();
        }
```

catsbi committed:
refactor: Lombok에 의해 생성된 보장된 메서드에 대한 테스트 제외 설정 추가 및 불필요 메서드 제거

catsbi committed:
fix: 삭제 실패 테스트 케이스에서 findAccount 테스트하는 오류 수정

catsbi commented:
Lombok에 의해 생성되는 메서드에 대해서는 getter, setter까진 몰라도 Builder의 ToString까지 테스트하기는 비용낭비인것 같아 Lombok.config를 통해 exclude로 제외해줬습니다. 

그밖에 코드 커버리지 리포트를 보면서 부족한 테스트 케이스 수정 및 삭제 테스트임에도 검색 테스트를 하는 오타? 수정해줬습니다 ㅎㅎ 서비스 영역의 메서드의 인자 및 반환타입 관련해서 DTO로 한 부분에 대해서는 바로 수정하기보단 좀 더 고민할 필요가 있어보여서 보류 했습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/dto/AccountDataTest.java:
```diff
@@ -0,0 +1,102 @@
+    private static ValidatorFactory factory;
+    private static Validator validator;
+
+
```
여기 빈 줄이 두 줄이 띄어져 있네요

hannut91 commented on app/src/test/java/com/codesoom/assignment/dto/AccountDataTest.java:
```diff
@@ -0,0 +1,102 @@
+        final Set<ConstraintViolation<AccountSaveData>> validate = validator.validate(accountData);
+
+        assertThat(validate).isEmpty();
+    }
```
`필드 검사를 할 수 있다`라고하면 결과가 어떻게 되어야하는지 예상하기 어려운 것 같아요. 왜 `validate`가 empty어야 하는지가 나타나지 않는 것 같습니다.

`validate는 유효한 객체의 경우에는 빈 Set을 반환합니다` 라고 하는 것은 어떨까요? 그럼 자연스럽게
`validate는 이름이 비어있으면 에러를 반환합니다`같은 것을 작성할 수 있을거에요

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/AccountTest.java:
```diff
@@ -0,0 +1,74 @@
+        assertThat(account.getName()).isEqualTo(ACCOUNT_NAME);
+        assertThat(account.getEmail()).isEqualTo(ACCOUNT_EMAIL);
+        assertThat(account.getPassword()).isEqualTo(ACCOUNT_PASSWORD);
+    }
```
셋 모두 동일한 결과를 가져오는 것 같습니다. 아래 3개가 모두 같은 결과를 낸다고 테스트를 작성해봐도 괜찮을 것 같아요

```java
Account.builder()
                .id(1L)
                .name(ACCOUNT_NAME)
                .email(ACCOUNT_EMAIL)
                .password(ACCOUNT_PASSWORD)
                .build();
Account.of(
    1L, ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD);
Account.from(new AccountSaveData(1L, ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD));
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/AccountControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+    @DisplayName("회원 정보 등록 API를 호출해 회원 정보를 등록할 수 있다. - POST /user")
+    @Test
+    void createAccount() throws Exception {
+        String content = String.format(FORM, null, ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD);
```
요청을 위한 JSON data를 만들려고 하신 것 같아요. `toJSON`같은 함수로 만드는 것은 어떨까요?

```java
String toJSON(Account account) {
    return String.format(FROM, account.getId(), account.getName(), account.getEmail(), account.getPassword());
}

String content = toJSON(account.of(null, ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD));
```

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/AccountControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+    @DisplayName("등록된 회원의 정보를 수정할 수 있다. - PATCH /user/{id}")
+    @Test
+    void patchWithExistsAccount() throws Exception {
+        final AccountSaveData savedAccountData = accountService.creation(
+                AccountSaveData.of(ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD));
+
+        final String content = String.format(FORM,
+                savedAccountData.getId(),
+                OTHER_ACCOUNT_NAME,
+                OTHER_ACCOUNT_EMAIL,
+                OTHER_ACCOUNT_PASSWORD);
```
Service를 모킹하고 있다면, `creation`이 반환하는 것은 `account`이죠? 따라서 그냥 `account`를 사용하면 안되나요?

catsbi replied:
accountService는 받는 인자와 반환 타입 모두 AccountSaveData(or AccountUpdateData)로 DTO를 주고받습니다. 그렇기에 creation이 반환하는 것은 AccountSaveData입니다.  도메인을 서비스레이어 밖으로 내보내지 않아 도메인 모델 보호에 중점을 맞췄습니다.

catsbi committed:
refactor: view 역할이 없는 수정용 dto에 불필요한 애노테이션 제거

catsbi committed:
refactor: 동일한 내용을 반환하는 테스트케이스 통합

catsbi committed:
refactor: 공백 2line 제거 및 테스트 주석 변경

catsbi committed:
refactor: 상품 컨트롤러 CORS 애노테이션 추가

catsbi commented:
리뷰주신 내용들 위주로 리팩토링을 진행하고 코드 커버리지 , api test, e2e test를 통과하는걸 위주로 진행했습니다. 추가적으로 계층형 테스트나 테스트케이스를 더 진행하고싶었으나 오늘자로 회사에 퇴사얘기를 하면서 뭔가 정신없이 바빠서 많은 진도를 빼지는 못했네요 ㅠ

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceNestedTest.java:
```diff
@@ -0,0 +1,180 @@
+            @DisplayName("정상적으로 생성되어 반환된다.")
+            @Test
+            void createAccount() {
+                AccountSaveData source = AccountSaveData.of(ACCOUNT_NAME, ACCOUNT_EMAIL, ACCOUNT_PASSWORD);
```
이 source는 테스트에 필요한 데이터인 것 같아요. 위에서 미리 만들어놓은 후 사용하는 것은 어떨까요?

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceNestedTest.java:
```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    @Nested
+    @DisplayName("create 메서드는")
+    class Describe_create {
+        @Nested
+        @DisplayName("인자값이 정상인 경우")
+        class Context_without_data {
+
+            @DisplayName("정상적으로 생성되어 반환된다.")
+            @Test
+            void createAccount() {
```
`create는 account를 생성해서 반환합니다` 정도로 작성해도 좋을 것 같습니다. 정상인 경우를 기술한다면, 비정상인 경우도 같이 기술하면 좋을 것 같아요

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceNestedTest.java:
```diff
@@ -0,0 +1,180 @@
+            @Test
+            void patchAccount() {
+                final AccountUpdateData target = AccountUpdateData.of(OTHER_ACCOUNT_NAME, OTHER_ACCOUNT_EMAIL, OTHER_ACCOUNT_PASSWORD);
+                final AccountSaveData updatedData = accountService.patchAccount(Describe_patch.this.accountData.getId(), target);
```
```suggestion
                final AccountSaveData updatedData = accountService.patchAccount(accountData.getId(), target);
```

hannut91 reviewed (commented):
아니 이럴수가 퇴사를 하시는군요 ㅠ 좋은 이유로 퇴사를 하셨으면 좋겠네요

catsbi committed:
refactor: 리뷰내용 적용및, reformat code,  필드 유효성 검사 로직 추가

- 필드 유효성 위배 사용자 정의 예외 객체 생성
- dto의 entity 변환 메서드에 유효성 검증 로직 추가
- 테스트 코드 필요한 부분 통합 및 코드 수정

hannut91 commented on app/src/test/java/com/codesoom/assignment/domain/AccountTest.java:
```diff
@@ -0,0 +1,60 @@
+
+        assertThatAccountDataField(account1);
+        assertThatAccountDataField(account2);
+        assertThatAccountDataField(account3);
```
`@ParameterizedTest`를 사용해볼까 생각도 해봤지만, 테스트 대상이 드러나지 않을 것 같아서 별로인 것 같았고, 똑같은 줄을 3개쓰는 것이 별로인 것 같아서 다음과 같이 해봤어요
```java
Stream.of(account1, account2, account3).forEach((it) -> {
    assertThat(it.getId()).isEqualTo(1L);
    assertThat(it.getName()).isEqualTo(ACCOUNT_NAME);
    assertThat(it.getEmail()).isEqualTo(ACCOUNT_EMAIL);
    assertThat(it.getPassword()).isEqualTo(ACCOUNT_PASSWORD);
});
```

catsbi replied:
이것도 좋은 생각인 것 같습니다!

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AccountServiceNestedTest.java:
```diff
@@ -0,0 +1,223 @@
+
+        @Nested
+        @DisplayName("인자값이 정상인 경우")
+        class Context_with_invalid_data {
+            private AccountSaveData accountFromEmptyName;
+            private AccountSaveData accountFromEmptyEmail;
+            private AccountSaveData accountFromEmptyPassword;
+
+            @BeforeEach
+            void prepareCreation() {
+                accountFromEmptyName = AccountSaveData.of("", ACCOUNT_EMAIL, ACCOUNT_PASSWORD);
+                accountFromEmptyEmail = AccountSaveData.of(ACCOUNT_NAME, "", ACCOUNT_PASSWORD);
+                accountFromEmptyPassword = AccountSaveData.of(ACCOUNT_NAME, ACCOUNT_EMAIL, "");
+            }
+
+            @DisplayName("Account 를 생성해 반환합니다.")
+            @Test
+            void createAccount() {
+                final String causeMessage = "공백일 수 없습니다";
+
+                assertThatThrownBy(() -> accountService.creation(accountFromEmptyName))
+                        .isInstanceOf(AccountFieldValidException.class)
+                        .hasMessageContaining(String.format(AccountFieldValidException.DEFAULT_MESSAGE_FORM,
+                                "name",
+                                causeMessage));
+
+                assertThatThrownBy(() -> accountService.creation(accountFromEmptyEmail))
+                        .isInstanceOf(AccountFieldValidException.class)
+                        .hasMessageContaining(String.format(AccountFieldValidException.DEFAULT_MESSAGE_FORM,
+                                "email",
+                                causeMessage));
+
+                assertThatThrownBy(() -> accountService.creation(accountFromEmptyPassword))
+                        .isInstanceOf(AccountFieldValidException.class)
+                        .hasMessageContaining(String.format(AccountFieldValidException.DEFAULT_MESSAGE_FORM,
+                                "password",
+                                causeMessage));
+
+            }
+        }
```
```java
@Nested
@DisplayName("인자값이 정상인 경우")
class Context_with_invalid_data {
    final String causeMessage = "공백일 수 없습니다";

    @DisplayName("Account 를 생성해 반환합니다.")
    @Test
    void createAccount() {
        Stream.of(
                Pair.of(
                        "name",
                        AccountSaveData.of("", ACCOUNT_EMAIL, ACCOUNT_PASSWORD)
                ),
                Pair.of(
                        "email",
                        AccountSaveData.of(ACCOUNT_NAME, "", ACCOUNT_PASSWORD)
                ),
                Pair.of(
                        "password",
                        AccountSaveData.of(ACCOUNT_NAME, ACCOUNT_EMAIL, "")
                )
        ).forEach((it) -> {
            String name = it.getFirst();
            AccountSaveData accountData = it.getSecond();

            assertThatThrownBy(() -> accountService.creation(accountData))
                    .isInstanceOf(AccountFieldValidException.class)
                    .hasMessageContaining(
                            String.format(
                                    AccountFieldValidException.DEFAULT_MESSAGE_FORM,
                                    name,
                                    causeMessage
                            )
                    );
        });
    }
}
```

이렇게 해보았는데, 너무 복잡할까요? 그냥 Parameterized테스트를 쓸 껄 그랬네요

catsbi committed:
refactor: 회원 객체 생성 테스트 케이스 Stream API를 사용하도록 변경, AccountService의 회원 생성 테스트 ArgumentsSource를 사용하여 테스트 파라미터 주입받도록 변경

catsbi committed:
refactor: 코드 리팩터링

- Generated 애노테이션 필요 없는 곳 제거
- 불필요한 유효성 검증 로직 제거
- 유효성 검증기 생성자 private 로 변경

hannut91 commented:
회고를 보니 좋은 결심을 하신 것 같네요. 분명 좋은 결과가 있으실거에요 ㅎㅎ 

이번 한 주도 고생 많으셨습니다.
